### ISSUE TITLE
Incorrect WITH clause parsing

### ISSUE THREAD

**Issue by Aurora-yzj** [memgraph/memgraph#1818]
**Memgraph version**
2.15.0

**Environment**
Operating system: Ubuntu 22.04
Architecture: AMD64
Connect method: any connect method
Driver: any driver

**Describe the bug**
Memgraph wrongly outputs an empty query result with the label.

**To Reproduce**
1. Run the following query to insert the initial value:
```
CREATE (n0 :L14:L6:L16{k90 : -6020343600929744027, k40 : "FDQiPAV", k86 : "ERfcO", k97 : -938949366447903622, k85 : "ovwhSLZAP", k41 : true, k88 : "lY585", k99 : "ZONlveyi", k87 : 8743627198698289702, k98 : false, k35 : "UnPskxq", k89 : false, k37 : false, k36 : "PL0c7c", k39 : "lFSr7Ayy6", k38 : 7940196328277848464, k101 : "L26pjmJ", k100 : "cyZP4", k102 : "yrPovB", id : 1});
CREATE (n0 :L14{k90 : -7060347969001340006, k86 : "iX5GN", k85 : "xenyWQ", k88 : "czFsW", id : 2, k87 : 865003363209998952, k89 : true});
CREATE (n0 :L14:L6{k90 : 7502890094296041241, k40 : "Ky8F2", k86 : "LrXHN21", k85 : "myTXdBWsg", k41 : false, k88 : "KwnFAQ", k87 : -7161657525350621404, k35 : "I2dy5", k89 : true, k37 : true, k36 : "guoMlWi", k39 : "eABAucriQ", k38 : -2334082973896046325, id : 8});
CREATE (n0 :L14{k90 : 6371917648852764880, k86 : "EUBTLSajb", k85 : "j3X7cu", k88 : "z7lOsE", id : 11, k87 : -3658949507297496944, k89 : true});
MATCH (n0 {id : 1}), (n1 {id : 8}) MERGE(n0)-[r :T28{k370 : "363nssm", k372 : "vMr7l0", k371 : 669476107921837474, k373 : 4675179864053376404, k369 : "hbaHc", id : 20, k368 : 7994893128816991024}]->(n1);
MATCH (n0 {id : 2}), (n1 {id : 11}) MERGE(n0)-[r :T28{k370 : "bJDiPLop", k372 : "nujvpPMnk", k371 : 4740270152855389496, k373 : 8170260913811077215, k369 : "UsqvNaGf", id : 22, k368 : 764222251036993284}]->(n1);
```
2. Run the query:
```
MATCH (n0)-[r0]->(n1) WITH r0 ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id
```
Memgraph wrongly gives an empty result. 
![2024-03-14_17-04](https://github.com/memgraph/memgraph/assets/24768784/78b75f66-86ab-444a-854b-8bfdc8202d3a)
If we modify it to an equivalent query, adding n0 to WITH clause:
```
MATCH (n0)-[r0]->(n1) WITH n0,r0 ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id
```
It normally gives the result.
![2024-03-14_17-04_1](https://github.com/memgraph/memgraph/assets/24768784/a8102bed-48e0-4acc-ab85-808fb9ea278c)

**Expected behavior**
The WITH clause is handled incorrectly. Memgraph gives wrong results.

Thank you so much for your help!

**Verification Environment**
* Plain memgraph Docker image
* Architecture: x86

**Comment by katarinasupe** [comment]
Hi @Aurora-yzj, thank you for reporting this issue. Would you expect `Unbound variable: n0` in case:

`MATCH (n0)-[r0]->(n1) WITH r0 ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id`

When I run 
```
MATCH (n0)-[r0]->(n1)
WITH n0
ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id
```
I get `Unbound variable: r0.` and that's the reason why I'm asking that.

**Comment by Aurora-yzj** [comment]
> Hi @Aurora-yzj, thank you for reporting this issue. Would you expect `Unbound variable: n0` in case:
> 
> `MATCH (n0)-[r0]->(n1) WITH r0 ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id`
> 
> When I run
> 
> ```
> MATCH (n0)-[r0]->(n1)
> WITH n0
> ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id
> ```
> 
> I get `Unbound variable: r0.` and that's the reason why I'm asking that.
Thanks for your prompt reply!
Actually, the query I tried is:
```
MATCH (n0)-[r0]->(n1) WITH r0 ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id
```
Yours was "WITH n0", while mine was "WITH r0".
In this case, I think the query should normally give me a result. I have also tried this query in Neo4j, it returned a result normally. But with Memgraph, it returned an empty result.
If I revise the query into an equivalent one (since I am not actually using n0 in the following clauses, adding n0 to WITH won't cause any difference):
```
MATCH (n0)-[r0]->(n1) WITH n0,r0 ORDER BY (r0.k372) DESC WHERE ("1" <> n0.k99) return r0.id
```
It then gave the result normally.
Thanks for your help!

**Comment by katarinasupe** [comment]
Yes, I changed it to n0 on purpose to show how it acts differently if I forward one variable vs another. 
Thanks for the report, I'll label the issue and let the team handle it when it becomes a priority.

**Comment by Aurora-yzj** [comment]
OK, now I get it. Thanks!
